Improve version, version range, and version list syntax and behavior

haampie opened the pull request:
## Version types, parsing and printing

- The version classes have changed: `VersionBase` is removed, there is now a
  `ConcreteVersion` base class. `StandardVersion` and `GitVersion` both inherit
  from this.

- The public api (`Version`, `VersionRange`, `ver`) has changed a bit:
  1. `Version` produces either `StandardVersion` or `GitVersion` instances.
  2. `VersionRange` produces a `ClosedOpenRange`, but this shouldn't affect the user.
  3. `ver` produces any of `VersionList`, `ClosedOpenRange`, `StandardVersion`
     or `GitVersion`.

- No unexpected type promotion, so that the following is no longer an identity:
  `Version(x) != VersionRange(x, x)`.

- `VersionList.concrete` now returns a version if it contains only a single element
  subtyping `ConcreteVersion` (i.e. `StandardVersion(...)` or `GitVersion(...)`)

- In version lists, the parser turns `@x` into `VersionRange(x, x)` instead
  of `Version(x)`.

- The above also means that `ver("x")` produces a range, whereas
  `ver("=x")` produces a `StandardVersion`. The `=` is part of _VersionList_
  syntax.

- `VersionList.__str__` now outputs `=x.y.z` for specific version entries,
  and `x.y.z` as a short-hand for ranges `x.y.z:x.y.z`.

- `Spec.format` no longer aliases `{version}` to `{versions}`, but pulls the
  concrete version out of the list and prints that -- except when the list is
  is not concrete, then is falls back to `{versions}` to avoid a pedantic error.
  For projections of concrete specs, `{version}` should be used to render
  `1.2.3` instead of `=1.2.3` (which you would get with `{versions}`).
  The default `Spec` format string used in `Spec.__str__` now uses
  `{versions}` so that `str(Spec(string)) == string` holds.

## Changes to `GitVersion`

- `GitVersion` is a small wrapper around `StandardVersion` which enriches it
   with a git ref. It no longer inherits from it.

- `GitVersion` _always_ needs to be able to look up an associated Spack version
  if it was not assigned (yet). It throws a `VersionLookupError` whenever `ref_version`
  is accessed but it has no means to look up the ref; in the past Spack would
  not error and use the commit sha as a literal version, which was incorrect.
   
- `GitVersion` is never equal to `StandardVersion`, nor is satisfied by it. This
  is such that we don't lose transitivity. This fixes the following bug on `develop`
  where `git_version_a == standard_version == git_version_b` does not imply
  `git_version_a == git_version_b`. It also ensures equality always implies equal
  hash, which is also currently broken on develop; inclusion tests of a set of
  versions + git versions would behave differently from inclusion tests of a
  list of the same objects.

- The above means `ver("ref=1.2.3) != ver("=1.2.3")` could break packages that branch
  on specific versions, but that was brittle already, since the same happens with
  externals: `pkg@1.2.3-external` suffixes wouldn't be exactly equal either. Instead,
  those checks should be `x.satisfies("@1.2.3")` which works both for git versions and
  custom version suffixes.

- `GitVersion` from commit will now print as `<hash>=<version>` once the
  git ref is resolved to a spack version. This is for reliability -- version is frozen
  when added to the database and queried later. It also improves performance
  since there is no need to clone all repos of all git versions after `spack clean -m`
  is run and something queries the database, triggering version comparison, such
  as potentially reuse concretization.

- The "empty VerstionStrComponent trick" for `GitVerison` is dropped since it wasn't
  representable as a version string (by design). Instead, it's replaced by `git`,
  so you get `1.2.3.git.4` (which reads 4 commits after a tag 1.2.3). This means
  that there's an edge case for version schemes `1.1.1`, `1.1.1a`, since the
  generated git version `1.1.1.git.1` (1 commit after `1.1.1`) compares larger
  than `1.1.1a`, since `a < git` are compared as strings. This is currently a
  wont-fix edge case, but if really required, could be fixed by special casing
  the `git` string.

- Saved, concrete specs (database, lock file, ...) that only had a git sha as their
  version, but have no means to look the effective Spack version anymore, will
  now see their version mapped to `hash=develop`. Previously these specs
  would always have their sha literally interpreted as a version string (even when
  it _could_ be looked up). This only applies to databases, lock files and spec.json
  files created before Spack 0.20; after this PR, we always have a Spack version
  associated to the relevant GitVersion).

- Fixes a bug where previously `to_dict` / `from_dict` (de)serialization would not
  reattach the repo to the GitVersion, causing the git hash to be used as a literal
  (bogus) version instead of the resolved version. This was in particularly breaking
  version comparison in the build process on macOS/Windows.


## Installing or matching specific versions

- In the past, `spack install pkg@3.2` would install `pkg@=3.2` if it was a
  known specific version defined in the package, even when newer patch releases
  `3.2.1`, `3.2.2`, `...` were available. This behavior was only there because
  there was no syntax to distinguish between `3.2` and `3.2.1`. Since there is
  syntax for this now through `pkg@=3.2`, the old exact matching behavior is
  removed. This means that `spack install pkg@3.2` constrains the `pkg` version
  to the range `3.2`, and `spack install pkg@=3.2` constrains it to the specific
  version `3.2`.

- Also in directives such as `depends_on("pkg@2.3")` and their when
  conditions `conflicts("...", when="@2.3")` ranges are ranges, and specific
  version matches require `@=2.3.`.

- No matching version: in the case `pkg@3.2` matches nothing, concretization
  errors. However, if you run `spack install pkg@=3.2` and this version
  doesn't exist, Spack will define it; this allows you to install non-registered
  versions.

- For consistency, you can now do `%gcc@10` and let it match a configured
  `10.x.y` compiler. It errors when there is no matching compiler.
  In the past it was interpreted like a specific `gcc@=10` version, which
  would get bootstrapped.

- When compiler _bootstrapping_ is enabled, `%gcc@=10.2.0` can be used to
  bootstrap a specific compiler version.

## Other changes

- Externals, compilers, and develop spec definitions are backwards compatible.
  They are typically defined as `pkg@3.2.1` even though they should be
  saying `pkg@=3.2.1`. Spack now transforms `pkg@3` into `pkg@=3` in those cases.

- Finally, fix strictness of `version(...)` directive/declaration. It just does a simple
  type check, and now requires strings/integers. Floats are not allowed because
  they are ambiguous `str(3.10) == "3.1"`.

Closes #36183

alalazo commented on lib/spack/spack/version.py:
```diff
@@ -59,199 +40,214 @@
 iv_min_len = min(len(s) for s in infinity_versions)
 
 
-def coerce_versions(a, b):
-    """
-    Convert both a and b to the 'greatest' type between them, in this order:
-           VersionBase < GitVersion < VersionRange < VersionList
-    This is used to simplify comparison operations below so that we're always
-    comparing things that are of the same type.
-    """
```
It would be :100: to get rid of this!

haampie commented:
```
PKG-DIRECTIVES: 3 issues found
1. py-apache-beam: dependency on py-dill@0.3.1.1:0.3.1 cannot be satisfied by known versions of py-dill
    happening in /home/runner/work/spack/spack/var/spack/repos/builtin/packages/py-apache-beam/package.py
    known versions of py-dill are 0.3.6, 0.3.5.1, 0.3.4, 0.3.1, 0.2.9, 0.2.7, 0.2.6, 0.2.5, 0.2.4, 0.2.3, 0.2.2, 0.2.1, 0.2
2. py-cutadapt: dependency on py-dnaio@0.3.0:0.3 cannot be satisfied by known versions of py-dnaio
    happening in /home/runner/work/spack/spack/var/spack/repos/builtin/packages/py-cutadapt/package.py
    known versions of py-dnaio are 0.9.1, 0.4.2, 0.3
3. py-sphinx-tabs: dependency on py-docutils@0.16.0:0.16 cannot be satisfied by known versions of py-docutils
    happening in /home/runner/work/spack/spack/var/spack/repos/builtin/packages/py-sphinx-tabs/package.py
    known versions of py-docutils are 0.19, 0.18.1, 0.18, 0.17.1, 0.17, 0.16, 0.15.2, 0.14, 0.13.1, 0.12
```

looks correct to me

haampie committed:
Revamp versions

haampie committed:
fix bugs caught by package audit

haampie committed:
fix cvs test

haampie committed:
Add a regression test for the upperbound bug, and some others

haampie committed:
more testing

haampie committed:
fix circular import

haampie committed:
attach git lookup also in from_dict deserialization

haampie committed:
typo

haampie committed:
add a test for git version (de)serialization

haampie committed:
skip test on windows

haampie committed:
GitVersion: append serializable spack version

Turn `pkg@<hash>` into `pkg@<hash>=<version>` on `str(...)`.

This way there is no need to keep a git repo cache around for eternity.

haampie committed:
fix types

haampie committed:
Don't make __str__ err when no ref version can be looked up

haampie committed:
add a test, remove user_supplied_reference attribute

haampie committed:
fixes

haampie committed:
require str(GitVersion(...)) to produce =version if ref_lookup

haampie committed:
fix test

haampie committed:
bring back gitversions up_to, dotted, etc

haampie committed:
fix version boolness; dont have it trigger git lookup

alalazo commented:
So far this is improving the setup phase in a noticeable way:
[radiuss_versions.csv](https://github.com/spack/spack/files/11149331/radiuss_versions.csv) (e424fd49ad9dfd9926f21927f9e13a6ee5633973)
[radiuss_develop.csv](https://github.com/spack/spack/files/11149332/radiuss_develop.csv) (6a2f80e2c655cba38a4efbda862a7d7060291444)
[radiuss.txt](https://github.com/spack/spack/files/11149333/radiuss.txt)

Here's the plot:
![totals](https://user-images.githubusercontent.com/4199709/229817462-dbf96a2f-45c9-4595-988e-227f39b1b61a.png)

haampie committed:
disallow version(::float, ...)

trws commented on lib/spack/spack/test/concretize_requirements.py:
```diff
@@ -199,54 +224,51 @@ def test_requirement_adds_git_hash_version(
-    s2 = Spec("v@{0}".format(commits[1])).concretized()
-    assert s2.satisfies("@{0}".format(commits[1]))
-    assert s2.satisfies("@2.3")
+    # s2 = Spec(f"v@{commits[1]}").concretized()
+    # assert s2.satisfies("v@{commits[1]}")
+    # assert s2.satisfies("@2.3")
```
Intentionally commented out?

haampie replied:
Commented out because of how weird gitversion version declaration was handled in asp.py.

Somehow, `@hash=version` on the command line both defines `hash=version` and `version` as possible/declared versions. That's wrong... And similarly the situation with preferences is weird: a preferred version is added to the list of declared versions, but is not among "possible versions" (and lacks this weirdness of declaring two versions for gitversion).

With @tgamblin's comment that it would be better to remove the "exact match from command line" logic when you actually pass a range instead of a specific version, I think some weirdness can simply be removed from asp.py; I'll also remove `hash=bla` defines `bla` as a version and revive the test

haampie replied:
Okay, yeah: this particular instance of requirements is intentional, because `spack install v@<hash>` causes a lookup which is independent of concretization, so it will resolve to a different version than is required, causing a concretizationn error. However, `spack install v@2.3` is fine, since it's a range which includes a git commit which is required.

imho it's too involved to make `spack install v@<hash>` resolve its version either to a lookup _or_ to a user defined mapping in a requirement -- before you get to the requirement, the git ref is very likely already resolved to some tag from the git repo.

trws commented on lib/spack/spack/test/concretize_requirements.py:
```diff
@@ -255,33 +277,32 @@ def test_preference_adds_new_version(
+    # s3 = Spec(f"v@{commits[1]}").concretized()
+    # assert s3.satisfies(f"@{commits[1]}")
     # Note: this check will fail: the command-line spec version is preferred
     # assert s3.satisfies("@2.3")
```
Same here.

haampie replied:
This is revived

trws commented on lib/spack/spack/version.py:
```diff
@@ -335,387 +366,198 @@ def up_to(self, index):
-        return hash(str(self))
+    def __lt__(self, other):
+        if isinstance(other, GitVersion):
+            return (self.ref_version, self.ref) < (other.ref_version, other.ref)
```
What do we get from the inclusion of `self.ref` here? With it there, `git.main` and `git.<ref>` can't match even if they resolve to the same commit, is that desirable?

haampie replied:
I guess this ultimately comes from: `a == b` should imply `hash(a) == hash(b)`. We don't want `hash(GitVersion)` to resolve `ref_version` since it should be possible to create dicts/sets of them when they're not yet resolved (package.versions is in fact such a dict). And `hash(GitVersion)` must be constant, so `ref` is the only hashable property. That means `==` cannot discard `ref`. So, `GitVersion.__eq__` makes sense if it's implmented in terms of `(ref_version, ref)`. And then I guess `<` is just using the same comparator. (It also ensures there's a total order, but idk if that's actually necessary.)

tgamblin replied:
I think this is the right behavior, for the reasons @haampie  stated *and* because any ref version is essentially unresolved -- I think if you want it to match exactly it should be resolved to a real commit, since the branch *could* move.  I think it would be really hard to handle `main` on an installed package, presumably from the past, vs. `main` now, if we didn't handle it the way @haampie is here.

trws commented:
Initial parse might be a bit slower, translating to ClosedOpenRange rather than a closed/closed?  Or maybe there are a couple more string passes from checking for `=` in some places? I would expect the overall to be better, and @haampie and I were just discussing and I think going the extra step to make StandardVersion itself a ClosedOpenRange to get the full comparison perf boost should remain feasible after this is in, maybe also GitVersion but they're strange enough that's not going to entirely go away.

tgamblin commented:
> Exact matching logic: on develop the behavior of spack install pkg@3.1 would install an exact match of pkg@=3.1 if both version 3.1 and 3.1.1 were defined in the package. For backwards compatibility this behavior is preserved. In the future we could change this behavior since users can now distinguish between pkg@=3.1 and pkg@3.1.

I actually think this should prefer 3.1.1 -- any reason not to do that?  I don't think we guarantee backward compatibility at this level, but if there's a nasty consequence to this that I'm missing lemme know.

tgamblin commented:
Ok, a nasty consequence for anyone with unresolved git versions installed: I happen to have `zlib@04f42ceca40f73e2978b50e93806c2a18c1281fc` installed. 

On `develop`:
```console
(spackbook):spack> spack find zlib
-- darwin-monterey-m1 / apple-clang@14.0.0 ----------------------
zlib@1.2.13  zlib@1.2.13  zlib@04f42ceca40f73e2978b50e93806c2a18c1281fc
```

On this branch:
```console
(spackbook):spack> spack find zlib
==> Error: git ref '04f42ceca40f73e2978b50e93806c2a18c1281fc' cannot be looked up: call attach_git_lookup_from_package first
```

Is there some way we can grandfather in installations like this?  I think it would be ok if we treated them like `@git.ref=develop` to avoid an extra lookup.

tgamblin commented on lib/spack/spack/build_systems/python.py:
```diff
@@ -290,7 +290,7 @@ def get_external_python_for_prefix(self):
         python_external_config = spack.config.get("packages:python:externals", [])
         python_externals_configured = [
-            spack.spec.Spec(item["spec"])
+            spack.spec.concrete_spec_from_old_syntax(item["spec"])
```
I think it's good to preserve backward compatibility here, but I wonder if we eventually want to make this and develop specs select a satisfying, known concrete version.

tgamblin commented on lib/spack/spack/solver/asp.py:
```diff
@@ -1601,48 +1605,47 @@ def key_fn(item):
             for idx, v in enumerate(version_preferences):
                 # v can be a string so force it into an actual version for comparisons
-                ver = spack.version.Version(v)
+                ver = vn.Version(v)
```
Ok this is making me realize that I don't currently like how we treat versions in `packages.yaml`.  I actually think it ought to be a range, as well, unless you're very specific.  e.g., IMO:

```console
spack -c "packages:hdf5:version:[1.18]" install hdf5
```

should get me the latest 1.18 version of `hdf5`.  i.e., the version there should be used in `satisfies()` to change the preference order of versions, not to add a new version as it does now.

We should probably keep the behavior for now and do a follow-on PR.  I'm trying to decide whether it also makes sense for externals and develop versions to do this.  Personally, I think it does since if you *really* mean some new version unknown to Spack, you can now specify `@=version`.  That would also AFAICT eliminate the need for `concrete_spec_from_old_syntax`.

tgamblin commented on lib/spack/spack/spec.py:
```diff
@@ -4799,6 +4804,23 @@ def __hash__(self):
+    interpreted_version = s.versions.concrete_range_as_version
+    if interpreted_version:
+        s.versions = vn.VersionList([interpreted_version])
+    return s
```
I wonder if this is clearer and easier to add when necessary:

```
spack.spec.Spec("foo@1.0", concrete_version=True)
```

I also wonder if we'll end up needing this function in the long run, so it may not be worth it.

haampie replied:
I hope we can delete it rather sooner than later :p

tgamblin commented on lib/spack/spack/spec.py:
```diff
@@ -4799,6 +4804,23 @@ def __hash__(self):
+def concrete_spec_from_old_syntax(string: str):
+    """Same as Spec(string), but interprets @x as @=x"""
+    s = Spec(string)
+    interpreted_version = s.versions.concrete_range_as_version
```
Does this fail if there isn't a version for some reason?  Should it?

haampie replied:
`versions` is always initialized as a VersionList, should be fine

tgamblin commented on lib/spack/spack/cmd/checksum.py:
```diff
@@ -83,22 +83,17 @@ def checksum(parser, args):
+    if not args.versions and args.preferred:
         versions = [preferred_version(pkg)]
+    else:
+        versions = list(map(Version, args.versions))
```
```suggestion
        versions = [Version(v) for v in args.versions]
```

I just find this more readable even though this isn't an awful place to use `map`.

tgamblin commented on lib/spack/spack/compilers/__init__.py:
```diff
@@ -67,7 +68,7 @@ def pkg_spec_for_compiler(cspec):
     else:
         spec_str = str(cspec)
-    return spack.spec.Spec(spec_str)
+    return spack.spec.concrete_spec_from_old_syntax(spec_str)
```
It wouldn't be awful for this to be a range as well.  It would allow us to get the latest patch version of the compiler, and would be less prickly.

tgamblin commented on lib/spack/spack/compilers/aocc.py:
```diff
@@ -143,5 +143,5 @@ def fflags(self):
         # This is a known issue for AOCC 3.0 see:
         # https://developer.amd.com/wp-content/resources/AOCC-3.0-Install-Guide.pdf
-        if self.real_version == ver("3.0.0"):
+        if self.real_version == ver("=3.0.0"):
```
why not convert to `.satisfies()` for this and other cases?

haampie replied:
real_version has no promotion from string in satisfies

haampie replied:
This is fixed

tgamblin commented on lib/spack/spack/compilers/apple_clang.py:
```diff
@@ -41,46 +41,46 @@ def extract_version_from_output(cls, output):
     def cxx11_flag(self):
         # Spack's AppleClang detection only valid from Xcode >= 4.6
-        if self.real_version < spack.version.ver("4.0"):
+        if self.real_version < spack.version.ver("=4.0"):
```
Isn't this fine as-is?  Is `@=` strictly necessary here?

haampie replied:
Yeah, should be, but I think comparison between range and exact version is more for internal use (just s.t. versionlist can use insertion sort). I think it may be suprising/confusing to people that  `1.5 > 1.0:2.0` is true.

tgamblin replied:
so wouldn't it be less of a footgun if VersionList's sort used some internal comparator, and the user-exposed operators compared based on the ends of the range?

haampie replied:
In principle it could use an internal comparator, but I think `<` should then just error for mixed Version / ClosedOpenRange since there's no natural total order (if you want 1 < 2:3 and 2:3 < 4, how do 2.5 and 2:3 compare)

tgamblin commented on lib/spack/spack/compilers/cce.py:
```diff
@@ -58,7 +58,7 @@ def link_paths(self):
     def is_clang_based(self):
         version = self._real_version or self.version
-        return version >= ver("9.0") and "classic" not in str(version)
+        return version >= ver("=9.0") and "classic" not in str(version)
```
These comparisons are making me want to add `@<=9.0`, `@<9.0`, `@>=9.0` and `@>9.0`, etc., but that can be for some other PR.

tgamblin commented on lib/spack/spack/cray_manifest.py:
```diff
@@ -74,13 +78,13 @@ def spec_from_entry(entry):
         )
 
-    spec_format = "{name}@{version} {compiler} {arch}"
+    spec_format = "{name}@={version} {compiler} {arch}"
```
I think I like the default range, but this does seem a little tricky.  Basically any place we read things in we have to be careful to interpret the version concretely.

This is actually not unlike how Specs were originally treated in Spack -- Specs read from files (like the database) had to be interpreted as implicitly concrete before we added the `concrete` attribute to them.  It defaults to `True` so you don't see it in the DB, but if you output an abstract spec it'll have `concrete=False` in it.

I am not sure how to avoid the need to be careful about this so I guess I'll just leave this as a comment.

haampie replied:
Agreed, also when you copy a literal string from `spack spec` output and feed it back into `spack spec`, it might concretize to a "more specific" version. Alternatively `str(StandardVersion)` could print as `=<version>`, but I didn't go for that because it felt like a big change.

Fortunately spec -> json -> spec is fine, because VersionList serializes as `versions` or `version` depending on whether it's concrete.

tgamblin commented on lib/spack/spack/directives.py:
```diff
@@ -341,16 +347,28 @@ def _execute_version(pkg):
             kwargs["checksum"] = checksum
 
-        # Store kwargs for the package to later with a fetch_strategy.
+        if isinstance(ver, float):
+            raise VersionError(
+                f"{pkg.name}: declared version '{ver}' in package should be a string. "
+                "floating point numbers are ambiguous."
+            )
+        elif not isinstance(ver, (int, str)):
+            raise VersionError(
+                f"{pkg.name}: declared version '{ver!r}' in package should be a string or int."
+            )
```
I feel like we should push people towards strings here, as ints can become floats easily with casual editing.  Right now one message says strings and the other says strings OR ints.

I see there are some packages that call `Version(int)` but I'd rather fix that in `develop`, deprecate ints entirely, and remove the int option in 0.21.

haampie replied:
I agree, but I don't think `int` could ever go wrong (we actually go from str -> int during parsing). The `str(3.10) -> "3.1"` is most serious, and can't be prevented. The error message could of course just mention strings to nudge folks towards that.

tgamblin replied:
Ok fair enough but let’s make them both “string or int” then

haampie replied:
That's done

tgamblin commented on lib/spack/spack/solver/asp.py:
```diff
@@ -1843,22 +1846,34 @@ def generate_possible_compilers(self, specs):
-                    compilers.append(
-                        compiler_cls(
-                            s.compiler, operating_system=None, target=None, paths=[None] * 4
+                # This is only for backward compatibility, where old spack
+                # would interpret %gcc@10.0 as %gcc@=10.0 when bootstrapping
+                # compilers.
+                compiler_version = s.compiler.versions.concrete_range_as_version
```
I'd be ok with changing this.  But it should probably be in another PR and should go in the 0.20 changelog.

haampie replied:
This is now changed such that it `spack install pkg@1.2.3` and `spack install pkg %compiler@1.2.3` behave the same.

They fail if `1.2.3` does not match a known package / compiler version.

For defining the version on the fly, you need `@=`.

tgamblin commented on lib/spack/spack/version.py:
```diff
@@ -335,387 +366,198 @@ def up_to(self, index):
+            return self._ref_version
+
+        if self.ref_lookup is None:
+            raise RuntimeError(
+                f"git ref '{self.ref}' cannot be looked up: "
+                "call attach_git_lookup_from_package first"
+            )
```
This should be an assertion rather than a runtime error, but see my note above on existing installations with unresolved git versions.  We need some type of solution for that other than "reinstall everything"

haampie replied:
I made it into a more specific error so it can be caught, which is necessary in some places

tgamblin reviewed (changes requested):
Here's a partial review -- I'll finish the complete review later but there are already some actionable things here.

haampie committed:
map -> list comprehension

haampie committed:
nudge users towards strings in version(...) directive

haampie committed:
Warn about unresolvable git sha versions in db

Specs that are in the database just with pkg@<hash> where the commit
hash is no longer resolvable (for example when the git attribute is
removed from package.py) now issue a warning when constructing the
database.

This sanity check is only necessary for old installations, since spack
from now on only adds specs to the database with resolved spack version.

haampie committed:
actually use StandardVersion for unresolvable GitVersion, since that doesnt break reuse

haampie commented:
Okay, two things: 

1. `from_node_dict` was another instance where the git repo was simply not attached after constructing a Spec, which is now fixed, so specs with just shas in the db can now resolve to their spack versions
2. whenever the database contains a spec with git sha version w/o associated spack version, that's unresolvable due to package.py dropping the git attribute or so, spack now issues a warning and recommends to remove it. the git versions gets replaced by a normal version; this is what it's like on develop right now, and the reason to not assign `=develop` is because it mutates the immutable spec and would lead to weird issues in reuse concretization.

haampie committed:
remove comment, also remove hasattr on spec instead of package

haampie committed:
RuntimeError -> VersionLookupError

haampie committed:
test without invoking version(...), but the wrapped internal thing

haampie committed:
style

haampie committed:
use literal input version

tgamblin commented:
I removed my zlib version earlier but I still have a one-off `clingo` commit in the DB.  I get this when doing `spack find` now:

```console
(spackbook):spack> spack find
fatal: repository 'foo' does not exist
==> Error: Invalid record in Spack database: hash: ebfvstdloypshiz37hhigps7pct3nzit, cause: ProcessError: Command exited with status 128:
    '/Users/gamblin2/src/spack/var/spack/environments/default/.spack-env/view/bin/git' '-c' 'advice.detachedHead=false' 'clone' '--bare' '--quiet' 'foo' '/Users/gamblin2/Workspace/src/spack/lib/spack/spack/foo'
/Users/gamblin2/Workspace/src/spack/opt/spack/.spack-db/index.json
```

haampie committed:
str or int

tgamblin commented:
Ok so part of this may be outside the scope of this particular PR, but it's all nasty and probably needs fixing.

More on my `zlib@commit`: I was playing wth git versions and added a `git` property to `zlib`, then installed `zlib@37ed2112a1727293f4fd431df85dc821ee475d3d`.  I've since *removed* the git URL for zlib -- I didn't bother upstreaming it though I do think we should add git URLs to more things.

Since my last comment above, I've *removed* that `zlib` installation, but it's still in the DB:

```console
(spackbook):spack> spack find zlib
-- darwin-monterey-m1 / apple-clang@14.0.0 ----------------------
zlib@1.2.13  zlib@1.2.13

-- darwin-ventura-m1 / apple-clang@14.0.0 -----------------------
zlib@1.2.13
==> 3 installed packages
```

```console
(spackbook):spack> spack find -m zlib
-- darwin-bigsur-skylake / apple-clang@13.0.0 -------------------
zlib@37ed2112a1727293f4fd431df85dc821ee475d3d

-- darwin-monterey-m1 / apple-clang@14.0.0 ----------------------
zlib@1.2.13  zlib@1.2.13

-- darwin-ventura-m1 / apple-clang@14.0.0 -----------------------
zlib@1.2.13
==> 4 installed packages
```

On this branch I can't run any find command:
```console
(spackbook):spack> spack find
==> Error: Invalid record in Spack database: hash: ebfvstdloypshiz37hhigps7pct3nzit, cause: AttributeError: type object 'Zlib' has no attribute 'git'
/Users/gamblin2/Workspace/src/spack/opt/spack/.spack-db/index.json
```

It looks like it's taking the unresolved git commit out of the DB and attempting to resolve it.  Basically I think that shouldn't ever happen -- but it seemed like that's what you thought too:

> whenever the database contains a spec with unresolvable git sha version w/o associated spack version (due to package.py dropping the git attribute or so), spack now issues a warning and recommends to remove it.

Is it attempting to *resolve* the SHA?  I think it shouldn't even do that.

I don't really agree with this:

> the reason to not assign `=develop` is because it mutates the otherwise immutable spec, and would lead to weird issues in reuse concretization/hashing. 

Concrete hashes in Spack are assigned at concretization time and are respected forever.  So you can still respect the identifier for the installation (the dag hash) while changing how a version on that spec is compared.  It should never mutate the hash, and in reuse concretization the hashes are independent of the spec contents -- you *can* tell the solver to treat the version like it's `=develop` without changing the hash or altering reuse.  Every hash spec in reuse currently implies its dependencies (by hash) so you won't run into any issues with consistency in the sub-DAG rooted at the unresolved-git-version spec.  It'll only affect how constraints from *dependents* are evaluated.

So I still think the thing to do is to read the git commit version from the database, keep it around, and have it compare as `=develop` or `=inf` or something.  I guess the question is where that transformation should happen.  I *think* it's fine to do on read, though it would be nice if the spec serialized to the same JSON as before, but it *necessarily* doesn't have to -- as long as the hash doesn't change I think it's fine.

tgamblin commented:
This is what I see:

```
  File "/Users/gamblin2/Workspace/src/spack/lib/spack/spack/cmd/find.py", line 248, in find
    results = args.specs(**q_args)
              ^^^^^^^^^^^^^^^^^^^^
  File "/Users/gamblin2/Workspace/src/spack/lib/spack/spack/cmd/common/arguments.py", line 85, in _specs
    return spack.store.db.query(**kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/gamblin2/Workspace/src/spack/lib/spack/spack/database.py", line 1584, in query
    local_results = set(self.query_local(*args, **kwargs))
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/gamblin2/Workspace/src/spack/lib/spack/spack/database.py", line 1568, in query_local
    with self.read_transaction():
  File "/Users/gamblin2/Workspace/src/spack/lib/spack/llnl/util/lock.py", line 721, in __enter__
    self._as = self._acquire_fn()
               ^^^^^^^^^^^^^^^^^^
  File "/Users/gamblin2/Workspace/src/spack/lib/spack/spack/database.py", line 1091, in _read
    self._read_from_file(self._index_path)
  File "/Users/gamblin2/Workspace/src/spack/lib/spack/spack/database.py", line 873, in _read_from_file
    invalid_record(hash_key, e)
  File "/Users/gamblin2/Workspace/src/spack/lib/spack/spack/database.py", line 845, in invalid_record
    raise CorruptDatabaseError(msg, self._index_path)
spack.database.CorruptDatabaseError: Invalid record in Spack database: hash: ebfvstdloypshiz37hhigps7pct3nzit, cause: AttributeError: type object 'Zlib' has no attribute 'git'
    /Users/gamblin2/Workspace/src/spack/opt/spack/.spack-db/index.json
```

It's coming from an attempt to get the `git` attribute from `zlib`, but there is not one.  I don't think it should even need that attribute.

haampie committed:
fix str(GitVersion) & improve backtraces of corrupt db errors

haampie committed:
map to =develop again

haampie committed:
Don't bother with stage dirs for things that are already installed

spec@<hash> as a _dependency_ caused the installer to attempt to create
a stage dir, which requires it to have a git property (which may not
exist if the package has removed it since it was installed).

haampie commented:
For some reason Spack would create a stage dir for dependencies that are already installed, which causes `pkg.git` to be referenced if the version is a GitVersion, which made `spack install --reuse` error when reusing git-version-specs who's packages no longer have a git attribute.

hopefully edge case driven development can stop here...

tgamblin commented on lib/spack/spack/database.py:
```diff
@@ -695,8 +695,32 @@ def _read_spec_from_dict(self, spec_reader, hash_key, installs, hash=ht.dag_hash
 
+    def _validate_version(self, spec):
+        # Specs installed from a git sha without an associated spack version, that cannot
+        # be resolved anymore because the repo has been removed from its package.py, get
+        # replaced by a literal string version, to avoid errors in comparison.
```
this isn't true anymore right?  We add `=develop`

tgamblin commented on lib/spack/spack/database.py:
```diff
@@ -695,8 +695,32 @@ def _read_spec_from_dict(self, spec_reader, hash_key, installs, hash=ht.dag_hash
+        # Specs installed from a git sha without an associated spack version, that cannot
+        # be resolved anymore because the repo has been removed from its package.py, get
+        # replaced by a literal string version, to avoid errors in comparison.
+        # TODO: Remove this ad-hoc logic in Spack 0.21. It's mostly there so that these
+        # spack-versionless specs don't break basic operations. Since Spack 0.20, the
+        # resolved version is stored in the DB, so this is not needed for newer installs.
```
I'm hesitant to remove this for a while because it's not just a feature that'll be disabled -- it'll break operations on existing installs w/o a good workaround.  We've maintained good backward compatibility for this particular scenario for a long time -- spack can read *very* old databases (b/c people keep the installs around).

haampie replied:
Agreed, that's a good thing. I'll drop the comment.

tgamblin commented on lib/spack/spack/database.py:
```diff
@@ -695,8 +695,32 @@ def _read_spec_from_dict(self, spec_reader, hash_key, installs, hash=ht.dag_hash
+
         return spec
 
+    def _validate_version(self, spec):
```
So, I like this fix, but I think it needs to apply in more places.  If we read a concrete spec from a lock file, a build cache, or an upstream, what happens?  Upstreams and build caches are handled here, I guess, but lockfiles?  Random `spec.json` files? (this last one I care less about)

trws replied:
FWIW, this can be handled much of the time in Spec, the class. That applies in the vast majority of places, I *think* this is a workaround for one of the few places that isn't enough.

haampie replied:
I've moved it into `Spec._mark_root_concrete`; it should work for specs from the DB, lock files, and spec.json files.

tgamblin commented on lib/spack/spack/directives.py:
```diff
@@ -330,30 +336,35 @@ def version(ver, checksum=None, **kwargs):
-        pkg.versions[version] = kwargs
-
-    return _execute_version
+def _execute_version(pkg, ver: Union[str, int], checksum=None, **kwargs):
```
Any particular reason not to keep this as an internal closure of `version()`?  Oh I see this is to make it more testable?  That seems like a good reason.

haampie replied:
Yeah, I just wanted to pass a `pkg` as namedtuple so that it all fits in a single test w/o a new mock package

tgamblin commented on lib/spack/spack/spec.py:
```diff
@@ -2291,12 +2291,17 @@ def from_dict(data):
-            return SpecfileV2.load(data)
-        return SpecfileV3.load(data)
+        # Any git version should
+        for s in spec.traverse():
+            s.attach_git_version_lookup()
```
Should _validate_version from the database really live here, right after we read specs in from some serialized format?  I think this covers lock files, build caches, upstreams, etc.  I also think it avoids in-memory parses, CLI stuff, and spec literals in YAML configs.

tgamblin commented on lib/spack/spack/test/cmd/env.py:
```diff
@@ -760,7 +760,7 @@ def packages_file(tmpdir):
 packages:
   mpileaks:
-    version: [2.2]
+    version: ["2.2"]
```
ah ok I see where we were leaking floats into the version constructor.

haampie replied:
Oof yeah. I think this actually warrants a config `update(...)` where `int/float` -> `str`. Pretty sure it will confuse folks...

trws replied:
Yeah, it should really be int, or str required.  The one test package.py file that had explicit floats was fixed too right?

tgamblin commented on lib/spack/spack/test/compilers/basics.py:
```diff
@@ -338,83 +337,83 @@ def test_default_flags():
-    supported_flag_test("f77_pic_flag", "-fPIC", "arm@1.0")
-    supported_flag_test("fc_pic_flag", "-fPIC", "arm@1.0")
-    supported_flag_test("opt_flags", ["-O", "-O0", "-O1", "-O2", "-O3", "-Ofast"], "arm@1.0")
+    supported_flag_test("openmp_flag", "-fopenmp", "arm@=1.0")
+    supported_flag_test("cxx11_flag", "-std=c++11", "arm@=1.0")
+    supported_flag_test("cxx14_flag", "-std=c++14", "arm@=1.0")
+    supported_flag_test("cxx17_flag", "-std=c++1z", "arm@=1.0")
+    supported_flag_test("c99_flag", "-std=c99", "arm@=1.0")
+    supported_flag_test("c11_flag", "-std=c11", "arm@=1.0")
+    supported_flag_test("cc_pic_flag", "-fPIC", "arm@=1.0")
+    supported_flag_test("cxx_pic_flag", "-fPIC", "arm@=1.0")
+    supported_flag_test("f77_pic_flag", "-fPIC", "arm@=1.0")
+    supported_flag_test("fc_pic_flag", "-fPIC", "arm@=1.0")
+    supported_flag_test("opt_flags", ["-O", "-O0", "-O1", "-O2", "-O3", "-Ofast"], "arm@=1.0")
```
Not 100% understanding why all these need to change.  Do they? Can't you just change supported_flag_test to use satisfies?

haampie replied:
Let me check the logic again. In some cases a compiler is looked up, in some cases a compiler is defined. If it's looked up in config, `@1.0` should be enough; if it's defined, `@=1.0` makes most sense.

tgamblin replied:
My only concern would be when people have to write compiler specs in `compilers.yaml`.  I think if they *have* to write `@=` there to define a compiler, it's not great -- for the same reason requiring every `version()` directive to use `"=4.0"` wouldn't be so great.

tgamblin replied:
My only concern would be when people have to write compiler specs in `compilers.yaml`.  I think if they *have* to write `@=` there to define a compiler, it's not great -- for the same reason requiring every `version()` directive to use `"=4.0"` wouldn't be so great.

haampie replied:
I've reverted the change here. I guess I started fixing tests before I modified the from_dict function

tgamblin commented on lib/spack/spack/test/concretize.py:
```diff
@@ -766,16 +767,15 @@ def test_adjusting_default_target_based_on_compiler(
-        with pytest.raises(spack.concretize.UnavailableCompilerVersionError):
-            s = Spec("mpileaks %gcc@10.2")
-            s.concretize()
+        # The behavior here has changed since #8735. Now %gcc@10.2
+        # is an abstract compiler spec, and it should first find a
```
I think given this we should unmark the regression -- right @alalazo?

tgamblin commented on lib/spack/spack/test/concretize.py:
```diff
@@ -766,16 +767,15 @@ def test_adjusting_default_target_based_on_compiler(
+        # The behavior here has changed since #8735. Now %gcc@10.2
+        # is an abstract compiler spec, and it should first find a
+        # matching compiler gcc@=10.2.1
+        assert Spec("mpileaks %gcc@10.2").concretized().compiler == CompilerSpec("gcc@=10.2.1")
```
honestly this behavior is way better.

haampie replied:
yeah, it's consistent

tgamblin commented on lib/spack/spack/test/conftest.py:
```diff
@@ -123,7 +123,7 @@ def mock_git_version_info(git, tmpdir, override_git_repos_cache_path):
        o first commit
 
-    The repo consists of a single file, in which the Version._cmp representation
+    The repo consists of a single file, in which the Version.ref_version representation
```
```suggestion
    The repo consists of a single file, in which the Version._ref_version representation
```

tgamblin commented on lib/spack/spack/test/installer.py:
```diff
@@ -480,10 +483,12 @@ def test_bootstrapping_compilers_with_different_names_from_spec(
 
-            assert spack.spec.CompilerSpec("oneapi@22.2.0") in spack.compilers.all_compiler_specs()
+            assert (
+                spack.spec.CompilerSpec("oneapi@=22.2.0") in spack.compilers.all_compiler_specs()
```
This is making me think again that the externally facing comparison should do something a bit more intuitive, while `VersionList` can still use an internal comparator with the nice range properties you want -- does that make sense?

I can see people getting bitten by this, both for `CompilerSpec` and `Spec`

haampie replied:
Hm, as far as I can see, this is only fixed by reintroducing an equality `@22.2.0 == @=22.2.0`, which is what the PR is ultimately fixing :p it could also be written in terms of `any(s.satisfies(...))`, but even then I would write `@=22.2.0` because the test is looking for a specific version

haampie replied:
I think the `x in ys` test is rarely used in a way that's problematic.

The gist of the issue is, if the rhs contains concrete things, an abstract lhs doesn't compare equal; it's common to accidentally write an abstract version in the lhs, and it returns false w/o error, so easy to miss.

For Spec instances however, you rarely see `x in ys` tests, it's much more common to have `y.satisfies(x) for y in ys`, which has the desired effect whether `x` has version `@22.2.0` or `@=22.2.0`. I guess you either check for a subset of properties of the rhs (so no == there, only satisfies), or you happen to have a fully concrete spec around from the db/lockfile/etc and you want to test if it's exactly in another container in which case `x in ys` is fine.

So, sure, you can easily accidentally test for an abstract lhs in list of concrete items, but I don't think it's a good idea to guess what the intention of the user was and choose an interpretation accordingly, since that leads to bugs.

tgamblin replied:
> The gist of the issue is, if the rhs contains concrete things, an abstract lhs doesn't compare equal; it's common to accidentally write an abstract version in the lhs, and it returns false w/o error, so easy to miss.

Yes exactly

> I don't think it's a good idea to guess what the intention of the user was and choose an interpretation accordingly, since that leads to bugs.

I think I'm fine with this because we're talking about `dict.__contains__`, and I think that will only be used internally. The one that's would be an issue is if `Spec.__contains__` had this behavior, because it hasn't in the past.  It has always traversed the spec to see if anything in the DAG satisfied the argument.  It'll continue to work as it has before.

tgamblin replied:
> The gist of the issue is, if the rhs contains concrete things, an abstract lhs doesn't compare equal; it's common to accidentally write an abstract version in the lhs, and it returns false w/o error, so easy to miss.

Yes exactly

> I don't think it's a good idea to guess what the intention of the user was and choose an interpretation accordingly, since that leads to bugs.

I think I'm fine with this because we're talking about `dict.__contains__`, and I think that will only be used internally. The one that's would be an issue is if `Spec.__contains__` had this behavior, because it hasn't in the past.  It has always traversed the spec to see if anything in the DAG satisfied the argument.  It'll continue to work as it has before.

tgamblin commented on lib/spack/spack/test/modules/lmod.py:
```diff
@@ -59,10 +59,10 @@ def test_file_layout(self, compiler, provider, factory, module_configuration):
         # compilers
         # Check that specs listed as core_specs are transformed to "Core"
-        if compiler == "clang@3.3" or spec_string == "mpich@3.0.1":
+        if compiler == "clang@=3.3" or spec_string == "mpich@3.0.1":
```
could these just be `satisfies()`?  I'd almost rather force satisfies than make people remember that they need `@=` everywhere.

haampie replied:
`compiler` is a string here :p

tgamblin commented on lib/spack/spack/test/versions.py:
```diff
@@ -98,160 +106,160 @@ def check_union(expected, a, b):
 
 
 def test_string_prefix():
-    assert_ver_eq("xsdk-0.2.0", "xsdk-0.2.0")
-    assert_ver_lt("xsdk-0.2.0", "xsdk-0.3")
-    assert_ver_gt("xsdk-0.3", "xsdk-0.2.0")
+    assert_ver_eq("=xsdk-0.2.0", "=xsdk-0.2.0")
+    assert_ver_lt("=xsdk-0.2.0", "=xsdk-0.3")
+    assert_ver_gt("=xsdk-0.3", "=xsdk-0.2.0")
```
shouldn't we also test that the ranges compare equal?

haampie replied:
it's tested in `test_version_ranges`

tgamblin commented on lib/spack/spack/test/versions.py:
```diff
@@ -808,3 +801,165 @@ def test_git_versions_without_explicit_reference(
+    assert is_specific.intersects(upperbound) and is_range.intersects(upperbound)
+
+
+@pytest.mark.skipif(sys.platform == "win32", reason="Not supported on Windows (yet)")
```
why no windows?

haampie replied:
It failed, but forgot why. All other `mock_git_version_info` are marked skipif windows too

tgamblin commented on lib/spack/spack/test/versions.py:
```diff
@@ -808,3 +801,165 @@ def test_git_versions_without_explicit_reference(
+    assert spack.spec.Spec.from_dict(spec.to_dict()).satisfies("@1.0")
+
+
+@pytest.mark.skipif(sys.platform == "win32", reason="Not supported on Windows (yet)")
```
why no windows?

tgamblin commented on lib/spack/spack/test/web.py:
```diff
@@ -100,47 +100,47 @@ def test_spider_no_response(monkeypatch):
-    assert ver("2.0.0") in versions
-    assert ver("3.0") in versions
-    assert ver("4.5") in versions
+    assert ver("=0.0.0") in versions
+    assert ver("=1.0.0") in versions
+    assert ver("=2.0.0") in versions
+    assert ver("=3.0") in versions
+    assert ver("=4.5") in versions
```
its making me uneasy that every use of `in` is going to need an `=`.  I think we should talk about this one and figure out if this one can be fixed... I'm not seeing an easy way though... maybe it's fine.

Can't these be changed to `Version()` to make them more clear?  Maybe the issue I'm having here is that the tests use `ver` all over the place and we should just be using `Version` for concrete stuff.

haampie replied:
Directly using `Version(...)` sounds good to me, it's clear

tgamblin reviewed (changes requested):
ok still making my way through this but I have more questions.

haampie committed:
drop comment and reduce noise from pkg@hash in db messages

haampie committed:
fix old comment

haampie committed:
remove regression mark, update comment in test

haampie committed:
fix comment

haampie committed:
ver -> Version

haampie committed:
on-the-fly defined versions should be @=1.2.3, not @1.2.3

spack install python@3.12 will not attempt to define 3.12 if nothing
matches, but spack install python@=3.12 does define it

Also remove the weirdness around GitVersion, where hash=version would
not only define hash=version, but also version.

Fix a bunch of tests that made up @versions.

haampie committed:
roundtrip Spec(str(Spec(...)))

haampie committed:
no gcc@10.2.0 in builtin.mock

haampie committed:
another @= instance

haampie committed:
more non-existing versions in tests

haampie committed:
yet more non-existing versions in tests

haampie committed:
fix requirements test

haampie committed:
fix test

haampie committed:
simplify

haampie committed:
make nicer

haampie committed:
simplify

haampie committed:
bring back color!

haampie committed:
fix spack spec output, and versions coloring

haampie committed:
fix printing

haampie committed:
even more issues

haampie committed:
fix tarball compiler formatting

haampie committed:
fix output test

haampie committed:
anoter test

haampie commented:
I believe the version yak is mostly shaved

haampie committed:
packages.yaml: update number -> str in pkg.versions list

trws commented:
Oh don't worry, I'll come back and continue the shaving, the performance shaving shall be done, and the yak shall be shorn...

Anyway, on this: 

> There were still some issues due to new pkg@=version syntax, such as spack add pkg@=1.0 adding actually pkg@1.0 because str(Spec(spec_str)) != spec_str.
> 
> The only easy fix I can think of is to make VersionList.__str__ print the relevant = for StandardVersion, while keeping StandardVersion.__str__ just print plain versions. So, = is really version list syntax, not version syntax.
> 
> That way, projections can use {version} (the concrete version), while the default Spec.__str__ an use {versions} (the list).

I'm not sure I love the difference, I take it a bunch of stuff fails if `StandardVersion` includes the `=`?  If so, maybe we could split it between `str` and `repr` so at least both meaningful representations are easily produced from the external interface?

haampie commented:
Hm, I actually think it makes a lot of sense if a version instance prints as plain `x.y.z`, since if you know you work with a specific version it's weird if it gets emphasized: f"Installing version {spec.version}" would look weird as "Installing version =1.2.3".

I don't think str/repr solves it, since repr is for generating Python code

trws commented:
I somewhat agree with that, but that emphasizes the oddness that is parsing that version as a range like we do.  Printing with the = means copy and pasting it is actually correct, so maybe we should just live with it?

-Tom (mobile)
________________________________
From: Harmen Stoppels ***@***.***>
Sent: Friday, April 7, 2023 11:59:52 AM
To: spack/spack ***@***.***>
Cc: Scogland, Tom ***@***.***>; Comment ***@***.***>
Subject: Re: [spack/spack] Fix version/versionrange/versionlist (PR #36273)


Hm, I actually think it makes a lot of sense if a version instance prints as plain x.y.z, since if you know you work with a specific version it's weird if it gets emphasized: f"Installing version {spec.version}" would look odd as "Installing version =1.2.3". Same in spack info etc.

I don't think str/repr solves it, since repr is for generating Python code

—
Reply to this email directly, view it on GitHub<https://urldefense.us/v3/__https://github.com/spack/spack/pull/36273*issuecomment-1500534961__;Iw!!G2kpM7uM-TzIFchu!jRZh-POHpIz26MZJikci0lOvExRGicIqd6Fjoyty4F27hfaTc_Xb1mkbwRLBT23vBQ$>, or unsubscribe<https://urldefense.us/v3/__https://github.com/notifications/unsubscribe-auth/AAFBFNI6ARRHP7GLV4M2TJDXABP2RANCNFSM6AAAAAAWCKFX4A__;!!G2kpM7uM-TzIFchu!jRZh-POHpIz26MZJikci0lOvExRGicIqd6Fjoyty4F27hfaTc_Xb1mkbwRLa_c3xUA$>.
You are receiving this because you commented.Message ID: ***@***.***>

haampie committed:
ensure associated spack version of git sha only specs on mark_concrete

haampie committed:
docs

tgamblin commented on lib/spack/docs/packaging_guide.rst:
```diff
@@ -892,6 +892,26 @@ use:
+which allows Spack to pick any ``3.x.y`` for Python, while
+``depends_on("python@=3.10.1")`` restricts it to a specific version.
+
+Only in exceptional cases do you need to specific versions,
```
```suggestion
Only in exceptional cases should you need to resort to specific `@=` versions,
```

tgamblin commented:
> so maybe we should just live with it?

I essentially implemented something like this years ago, with `@4.5.` for specific versions instead of `@=4.5`.  I stopped because printing the specific version *everywhere* was just too painful -- and it makes the interface *weird*.

If `spack find` prints `@=` everywhere, it's consistent in some sense, but it's also unnecessary and a waste of characters, because everyone knows those are specific versions.  They're *installed*.  Also, to someone who hasn't read the docs, it's just line noise, and I worry it'll also scare people away from Spack.  I think the way @haampie is proposing to do this makes a lot of sense and allows the exceptional thing to be used in the exceptional cases.  You *can* be specific when needed, but you don't have to be.  That is how the versions in Spack were intended -- they were meant to read they way people actually talk about versions.

trws commented:
I largely agree with that, but it does mean we have to be careful to maintain the split everywhere, and figure out which one makes sense.  You're right that when they're installed for example, they're concrete and users should know that, but for a new user who sees a package with say `1.1` and `1.1.5` and wants `1.1` what do you think they will type?  The result will likely either be that they end up selecting both or accidentally getting 1.1.5, which I suppose is what currently happens, but the version printed will not be sufficient to actually select the specific item.  There are enough places that I'm not sure which it should be that I was tempted to stick with one that would be consistent, but maybe it's worth it to reduce the visual noise, honestly I don't know.

tgamblin commented:
> for a new user who sees a package with say 1.1 and 1.1.5 and wants 1.1 what do you think they will type?

I’m ok if they type 1.1 and get 1.1.5 — I actually think it’s *better* that that is the default. If they’re confused, they can currently look this up. I think we could add a little info (not even warning) on load that says something like:

```
multiple matches; run ‘spack match foo@1.1’ for options
```

and have that show them the “minimal” specs for loading any variants (the spack diff logic can almost do this now).  That doesn’t exist yet but I think it’s better than `=` everywhere. It’s something people have wanted `spack load` and other commands to do when ambiguous — more descriptive specs instead of “pick a hash”. FWIW, pick a hash always works and would work here too, though I realize it’s another cognitive jump.

trws commented:
Agreed, and the minimal specs (even just having it print the options with short-refs as a first cut) would help a lot on commands like that.

tgamblin commented on lib/spack/spack/test/mirror.py:
```diff
@@ -201,7 +201,7 @@ def test_invalid_json_mirror_collection(invalid_json, error_message):
 
 def test_mirror_archive_paths_no_version(mock_packages, config, mock_archive):
-    spec = Spec("trivial-install-test-package@nonexistingversion").concretized()
+    spec = Spec("trivial-install-test-package@=nonexistingversion").concretized()
```
this doesn't have to change, right?

haampie committed:
add some missing tests/implementations

haampie committed:
More docs

haampie committed:
more docs

haampie committed:
style

haampie committed:
undo unrelated change

haampie committed:
deal with old concretizer ending up with non-concrete version

haampie committed:
pkg@x parsing also for compiler specs, simplify compiler logic, require %gcc@=10.1.0 for bootstrapping

haampie committed:
deal with compiler@=version bootstrapping tests

haampie committed:
use satisfies in aocc.py

haampie committed:
ver("=...") -> Version("...")

haampie committed:
define ver(":") in version.py

haampie committed:
remove exact_match logic

haampie committed:
two more instances of compiler bootstrapping

haampie committed:
more cases of vn.any_version

haampie committed:
style

haampie committed:
Fix the docs again

haampie committed:
improve stacktrace

haampie committed:
fix spec holding refernce to global any version

haampie commented:
I also removed the exact_match logic for `depends_on("pkg@2.3")` selecting `@2.3`. I believe everything is consistent now, you just need to specify `@=` for exact versions.

aweits commented:
This also fixes this behavior (in current develop):
```
[aweits@coltrane spack]$ spack-python 
Spack version 0.20.0.dev0
Python 3.9.14, Linux x86_64
>>> v1=spack.version.Version("git.5a7a9bd0e7295628f465c3bb4e42563b8b8c1a9c=1.7.0")
>>> v2=spack.version.Version("1.7.1")
>>> v2 > v1
False
>>> v2 < v1
True
```

haampie committed:
use compiler@version again

haampie commented:
@kwryankrattiger can you check if b8ad783c5e758a847abf07b3937b8a7c81577084 is correct?

haampie commented:
Looks like this is in good shape again

haampie commented:
@spackbot rebuild everything

haampie committed:
Fix use of ver in intel.py

haampie committed:
fix test

haampie committed:
style

haampie commented:
I benchmarked it once more with `spack solve --timers cp2k+sirius` on some hpc system:

Before:
```
    setup         20.348s
    load           0.037s
    ground         4.797s
    solve         11.239s
    total         37.111s
```

After:
```
    setup         14.898s
    load           0.035s
    ground         4.732s
    solve         11.229s
    total         32.122s
```

haampie committed:
also use {versions} for spack solve

alalazo commented:
Posting 2 benchmarks below. This gives around a 10% speedup on average for "fresh" concretizations, due to speed-ups in the setup phase.

## Benchmark 1: radiuss packages

[radiuss-versions.csv](https://github.com/spack/spack/files/11287045/radiuss-versions.csv)
[radiuss-develop.csv](https://github.com/spack/spack/files/11287046/radiuss-develop.csv)
[radiuss.txt](https://github.com/spack/spack/files/11287047/radiuss.txt)

![radiuss](https://user-images.githubusercontent.com/4199709/233391728-ecf9ebbb-8059-47bd-a967-e6e6b8b5ec5a.png)

## Benchmark 2: e4s packages

[e4s-versions.csv](https://github.com/spack/spack/files/11287100/e4s-versions.csv)
[e4s-develop.csv](https://github.com/spack/spack/files/11287101/e4s-develop.csv)
[e4s.txt](https://github.com/spack/spack/files/11287102/e4s.txt)
 
![e4s](https://user-images.githubusercontent.com/4199709/233393283-ec955c12-ab9f-44f2-9b46-5893451e6898.png)

<details>

<summary>Speedup table</summary>

| Spec solved      |   Total time [develop, secs.] |   Total time [PR, secs.] |   Speedup [%] |
|------------------|-------------------------------|--------------------------|---------------|
| adios2           |                    24.8122    |               22.8656    |       7.84514 |
| aml              |                    13.4705    |               12.2822    |       8.82154 |
| amrex            |                    26.4963    |               24.2166    |       8.60374 |
| aomp             |                    14.5789    |               13.2413    |       9.17516 |
| arborx           |                    24.4905    |               22.3969    |       8.54862 |
| archer           |                    14.3856    |               13.0198    |       9.4945  |
| argobots         |                    13.2301    |               11.7694    |      11.0413  |
| ascent           |                    33.247     |               31.4358    |       5.44758 |
| bolt             |                    14.1       |               13.1272    |       6.89941 |
| cabana           |                    25.4609    |               23.3437    |       8.31572 |
| caliper          |                    14.807     |               13.5975    |       8.16843 |
| chai             |                    15.0627    |               13.7715    |       8.57209 |
| charliecloud     |                    13.6959    |               12.8656    |       6.06251 |
| conduit          |                    15.4641    |               14.22      |       8.0455  |
| darshan-runtime  |                    14.4256    |               13.0472    |       9.55537 |
| darshan-util     |                    13.2254    |               11.7409    |      11.2252  |
| datatransferkit  |                    25.2451    |               23.0224    |       8.80426 |
| dyninst          |                    14.0862    |               12.9109    |       8.34316 |
| faodel           |                    14.2711    |               12.6902    |      11.0777  |
| flecsi           |                    17.1533    |               15.6476    |       8.77778 |
| flit             |                    13.8851    |               12.5924    |       9.30951 |
| flux-core        |                    14.9952    |               13.7105    |       8.56729 |
| flux-sched       |                    15.0142    |               13.4923    |      10.1362  |
| fortrilinos      |                    25.2936    |               22.9428    |       9.294   |
| gasnet           |                    13.8672    |               12.4145    |      10.4757  |
| geopm            |                    19.9702    |               18.3218    |       8.25424 |
| ginkgo           |                    15.9125    |               14.2115    |      10.6897  |
| globalarrays     |                    14.5543    |               13.1885    |       9.38392 |
| gotcha           |                    13.9693    |               12.3519    |      11.5786  |
| hdf5             |                    14.0342    |               12.6988    |       9.51504 |
| hdf5-vol-async   |                    14.2022    |               13.1284    |       7.56041 |
| heffte           |                    16.177     |               14.6712    |       9.30826 |
| hpctoolkit       |                    14.9625    |               13.5302    |       9.57266 |
| hpx              |                    15.9068    |               13.1389    |      17.4004  |
| hypre            |                    17.0162    |               15.0695    |      11.4405  |
| kokkos           |                    15.3845    |               13.3209    |      13.413   |
| kokkos-kernels   |                    15.7414    |               13.6491    |      13.2917  |
| legion           |                    14.9981    |               13.0887    |      12.7307  |
| libnrm           |                    14.0453    |               12.6313    |      10.0672  |
| libpressio       |                    23.7904    |               21.3248    |      10.364   |
| libpressio-tools |                    26.101     |               23.664     |       9.33698 |
| libquo           |                    14.2034    |               12.702     |      10.5711  |
| llvm             |                    13.8089    |               12.4324    |       9.96804 |
| llvm-doe         |                    14.1151    |               12.7914    |       9.3781  |
| loki             |                     0.0995932 |                0.111912  |     -12.3694  |
| mercury          |                    13.8379    |               12.5617    |       9.22291 |
| metall           |                    13.7732    |               12.5162    |       9.12675 |
| mfem             |                    27.9555    |               25.9556    |       7.15385 |
| mpark-variant    |                    13.5435    |               12.4286    |       8.23226 |
| mpich            |                    13.9382    |               12.4594    |      10.6097  |
| mpifileutils     |                    14.4101    |               13.0203    |       9.64466 |
| netlib-scalapack |                    14.1283    |               12.7504    |       9.75232 |
| ninja            |                    13.6935    |               12.3179    |      10.0462  |
| nvhpc            |                     0.12086   |                0.119332  |       1.2639  |
| omega-h          |                    26.3855    |               24.0391    |       8.89281 |
| openmpi          |                    13.9631    |               12.5053    |      10.4409  |
| openpmd-api      |                    25.2051    |               23.2739    |       7.66194 |
| papi             |                    12.7897    |               11.6279    |       9.08424 |
| papyrus          |                    14.1396    |               12.6398    |      10.6074  |
| parallel-netcdf  |                    14.0573    |               12.618     |      10.2388  |
| paraview         |                    27.9368    |               25.5044    |       8.70699 |
| parsec           |                    17.7739    |               16.0384    |       9.76477 |
| pdt              |                     0.114537  |                0.0940707 |      17.8684  |
| petsc            |                    23.8798    |               22.0507    |       7.65963 |
| phist            |                    23.7784    |               21.5843    |       9.22728 |
| plasma           |                    14.1195    |               12.632     |      10.5352  |
| precice          |                    25.2198    |               23.2204    |       7.9279  |
| pumi             |                    14.1052    |               12.972     |       8.03421 |
| py-cinemasci     |                    24.9154    |               22.1994    |      10.9011  |
| py-jupyterhub    |                    19.2633    |               17.2827    |      10.2814  |
| py-libensemble   |                    28.5162    |               26.0417    |       8.67754 |
| py-warpx         |                    38.8483    |               35.9437    |       7.47694 |
| qthreads         |                    14.7359    |               12.95      |      12.1191  |
| raja             |                    14.6189    |               13.5079    |       7.6001  |
| rempi            |                    14.4205    |               13.0633    |       9.41161 |
| scr              |                    15.4361    |               14.1215    |       8.5166  |
| slate            |                    15.9053    |               14.5358    |       8.61052 |
| slepc            |                    26.0067    |               24.188     |       6.99332 |
| stc              |                    15.1796    |               13.4798    |      11.1976  |
| strumpack        |                    18.0363    |               15.8376    |      12.1906  |
| sundials         |                    26.4385    |               23.1456    |      12.4551  |
| superlu          |                    14.6148    |               12.9838    |      11.16    |
| superlu-dist     |                    17.6314    |               14.3761    |      18.4633  |
| swig             |                    13.9299    |               11.9247    |      14.3947  |
| sz               |                    13.8862    |               12.4396    |      10.4178  |
| tasmanian        |                    16.1798    |               14.622     |       9.62818 |
| tau              |                    24.3037    |               22.4013    |       7.82764 |
| trilinos         |                    23.8908    |               22.1351    |       7.34886 |
| umap             |                    13.9273    |               12.6716    |       9.01573 |
| umpire           |                    14.2988    |               12.981     |       9.21675 |
| unifyfs          |                    14.4503    |               13.1428    |       9.0486  |
| upcxx            |                    14.4079    |               13.5915    |       5.66592 |
| veloc            |                    14.6353    |               13.0128    |      11.086   |
| vtk-m            |                    14.8506    |               13.5442    |       8.79681 |
| warpx            |                    35.05      |               32.5739    |       7.06455 |
| zfp              |                    14.0393    |               12.6585    |       9.835   |

</details>

haampie committed:
make mypy happy

tgamblin commented on lib/spack/docs/basic_usage.rst:
```diff
@@ -1121,36 +1136,35 @@ reference provided.  Acceptable syntaxes for this are:
-associated version is supplied by the user, Spack treats the git version as equivalent to that
-version for all version comparisons in the package logic (e.g. ``depends_on('foo', when='@1.5')``).
+If the git ref is not exactly a tag, then the distance to the nearest tag
+is also part of the resolved version. ``@git.abcdefg=0.19.git.8`` means
```
This is minor, and can be done in a (quick?) follow-on PR, but I think we could make this more readable with a different word, e.g. one of these:

```
git.abcdefg=0.19.plus.8
git.abcdefg=0.19.plus.8.commits
git.abcdefg=0.19.ahead.8
git.abcdefg=0.19.with.8
```

The redundant `git` doesn't offer a lot of intuition to someone reading this... I think `plus.8` is my favorite of the above options.

tgamblin commented on lib/spack/docs/packaging_guide.rst:
```diff
@@ -851,16 +851,16 @@ Version comparison
 
 Some Spack versions involve slight extensions of numeric syntax; for
-example, ``py-sphinx-rtd-theme@0.1.10a0``.  In this case, numbers are
+example, ``py-sphinx-rtd-theme@=0.1.10a0``.  In this case, numbers are
```
I think the introduction of `=` here doesn't add much and might be confusing, in that it draws your attention away from the `a`, which is the point of the example.  Is `=` needed?

tgamblin commented on lib/spack/spack/cmd/develop.py:
```diff
@@ -75,9 +79,12 @@ def develop(parser, args):
 
     spec = specs[0]
-    if not spec.versions.concrete:
+    version = spec.versions.concrete_range_as_version
```
consider renaming this to `as_concrete_version`

tgamblin commented on lib/spack/spack/version.py:
```diff
@@ -335,387 +366,205 @@ def up_to(self, index):
+        normalized_string = string[4:] if self.has_git_prefix else string
 
-        ref_lookup = self.ref_lookup or other_lookups
+        if "=" in normalized_string:
+            # Store the git reference, and parse the user provided version.
+            self.ref, spack_version = normalized_string.split("=")
+            self._ref_version = StandardVersion(
+                spack_version, *parse_string_components(spack_version)
+            )
+        else:
+            # The ref_version is lazily attached after parsing, since we don't know what
+            # package it applies to here.
+            self._ref_version = None
+            self.ref = normalized_string
```
I think this could be made a little clearer/faster by starting with:

```python
    self.ref, _, spack_version = normalized_string.partition("=")
```

but it's fine for now.

tgamblin reviewed (approved):
I have some comments that can be addressed in a follow-on, but this is ready to merge.

Thanks a lot for all the work on this @haampie!
